server/games/ladder_game.py

```diff
@@ -1,23 +1,66 @@
+import asyncio
 import logging
 from typing import Optional
 
 from server.config import config
 from server.players import Player
-from server.rating import RatingType
 
 from .game import Game
 from .game_results import ArmyOutcome, GameOutcome
-from .typedefs import FeaturedModType, GameType, InitMode
+from .typedefs import GameState, GameType, InitMode
 
 logger = logging.getLogger(__name__)
 
 
+class GameClosedError(Exception):
+    """
+    The game has been closed during the setup phase
+    """
+
+    def __init__(self, player: Player):
+        super().__init__()
+        self.player = player
+
+
 class LadderGame(Game):
     """Class for 1v1 ladder games"""
 
     init_mode = InitMode.AUTO_LOBBY
     game_type = GameType.MATCHMAKER
 
+    def __init__(self, id_, *args, **kwargs):
+        super().__init__(id_, *args, **kwargs)
+        self._launch_future = asyncio.Future()
+
+    async def wait_hosted(self, timeout: float):
+        return await asyncio.wait_for(
+            self._hosted_future,
+            timeout=timeout
+        )
+
+    async def wait_launched(self, timeout: float):
+        return await asyncio.wait_for(
+            self._launch_future,
+            timeout=timeout
+        )
+
+    async def launch(self):
+        await super().launch()
+        self._launch_future.set_result(None)
+
+    async def check_game_finish(self, player):
+        if not self._hosted_future.done() and (
+            self.state in (GameState.INITIALIZING, GameState.LOBBY)
```

Comment:
Shouldn't we be checking that `player == self.host` here?

Reply:
as long as the hosted future is not done the guests won't start connecting. So there is no way they could trigger this

Reply:
Then maybe add `assert self.host == player` to document this and prevent refactors from accidentally breaking this behavior.